yarpctest/recorder/recorder.go

```diff
@@ -0,0 +1,324 @@
+	}
+	return fmt.Sprintf("%x", hash.Sum64())
+}
+
+func (r *Recorder) makeFilePath(request *transport.Request, hash string) string {
+	s := fmt.Sprintf("%s.%s.%s.yaml", request.Service, request.Procedure, hash)
+	return filepath.Join(r.recordsDir, sanitizeFilename(s))
+}
+
+// Call implements the yarpc transport filter interface
+func (r *Recorder) Call(
+	ctx context.Context,
+	request *transport.Request,
+	out transport.Outbound,
+) (*transport.Response, error) {
+	log := r.logger
+	requestBody, err := ioutil.ReadAll(request.Body)
+	request.Body = bytes.NewReader(requestBody)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	requestHash := r.hashRequest(request, requestBody)
+	filepath := r.makeFilePath(request, requestHash)
+
+	switch r.mode {
+	case Replay:
+		return r.loadRecord(filepath)
+	case Append:
+		response, err := r.loadRecord(filepath)
+		if !os.IsNotExist(err) {
+			return response, err
+		}
+	}
+
+	response, err := out.Call(ctx, request)
+	if err != nil {
+		return response, err
+	}
+
+	if err := r.saveRecord(request, requestBody, response, filepath); err != nil {
+		log.Fatal(err)
+	}
+	return response, nil
+}
+
+// loadRecord attempts to load a record from the given file.
+func (r *Recorder) loadRecord(filepath string) (*transport.Response, error) {
+	// TODO wrap recorder errors in theirs own type.
+	rawRecord, err := ioutil.ReadFile(filepath)
+	if err != nil {
+		return nil, err
+	}
+	var cachedRecord record
+	if err := yaml.Unmarshal(rawRecord, &cachedRecord); err != nil {
+		return nil, err
+	}
+	response := transport.Response{
+		Headers: transport.HeadersFromMap(cachedRecord.Response.Headers),
+		Body:    ioutil.NopCloser(bytes.NewReader(cachedRecord.Response.Body)),
+	}
+	return &response, nil
+}
+
+// saveRecord attempts to save a record to the given file.
+func (r *Recorder) saveRecord(request *transport.Request, requestBody []byte,
+	response *transport.Response, filepath string) error {
+
+	responseBody, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return err
+	}
+	response.Body = ioutil.NopCloser(bytes.NewReader(responseBody))
+
+	if err := os.MkdirAll(recorderDir, 0775); err != nil {
+		return err
+	}
+
+	rawRecord, err := yaml.Marshal(&record{
+		Request: requestRecord{
+			Caller:    request.Caller,
+			Service:   request.Service,
+			Procedure: request.Procedure,
+			Encoding:  string(request.Encoding),
+			Headers:   request.Headers.Items(),
+			Body:      requestBody,
+		},
+		Response: responseRecord{
+			Headers: response.Headers.Items(),
+			Body:    responseBody,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(filepath, rawRecord, 0664)
+}
+
+type requestRecord struct {
+	Caller    string
+	Service   string
+	Procedure string
+	Encoding  string
+	Headers   map[string]string
+	Body      base64blob
+}
+
+type responseRecord struct {
+	Headers map[string]string
+	Body    base64blob
+}
+
+type record struct {
+	Request  requestRecord
+	Response responseRecord
+}
```

Comment:
Basically, I'm thinking that instead of `request` and `response` being top-level, you have a top-level `version` and `recording`:

```
version: 1
recording:
  request: ...
  response: ...
```